### PR TITLE
All API stubs for Smart Class Parameter Feature

### PR DESCRIPTION
**Pull request by jyejare**
All and Final set of API stubs for Smart Class Parameter Feature.
Includes below sub-components :
1. List
2. Override
3. Puppet Default
4.  Parameter Type
5. Optional Validation
6. Validation
7. Matcher
8.  Outside Impact
9. Value Hiding

Thanks !

### CODE REVIEW

File: tests/foreman/api/test_classparameters.py [hunk 1]
```diff
@@ -0,0 +1,1035 @@
...
+        2.  Error raised for invalid default value.
+
+        @status: Manual
+        """
+
+    @run_only_on('sat')
+    @stubbed()
+    @tier3
+    def test_positive_validate_puppet_default_value(self):
+        """Validation doesn't works on puppet default value.
```
Comment (oshtaier) [inline tests/foreman/api/test_classparameters.py]: positive - and `doesn't work`?
Even that is true, please correct `doesn't works` on `doesn't work`

File: tests/foreman/api/test_classparameters.py [hunk 2]
```diff
@@ -0,0 +1,1035 @@
...
+    @tier3
+    def test_negative_validate_matcher_value_with_regex(self):
+        """Error raised for matcher value not matching with regex.
+
+        @feature: Class Parameters - Optional Validation
+
+        @steps:
+
+        1.  Set override to True.
+        2.  Create a matcher with value that doesn't matches
```
Comment (oshtaier) [inline tests/foreman/api/test_classparameters.py]: `doesn't matches` - `doesn't match`

File: tests/foreman/api/test_classparameters.py [hunk 3]
```diff
@@ -0,0 +1,1035 @@
...
+        @assert: No error raised for non-empty default value
+
+        @status: Manual
+        """
+
+    @run_only_on('sat')
+    @stubbed()
+    @tier3
+    def test_negative_validate_matcher_value_required_checkbox(self):
+        """Error raised for blank matcher Value - Required check.
```
Comment (oshtaier) [inline tests/foreman/api/test_classparameters.py]: `Error is raised`

File: tests/foreman/api/test_classparameters.py [hunk 4]
```diff
@@ -0,0 +1,1035 @@
...
+        @assert: Error raised for blank matcher value.
+
+        @status: Manual
+        """
+
+    @run_only_on('sat')
+    @stubbed()
+    @tier3
+    def test_positive_validate_matcher_value_required_checkbox(self):
+        """Error not raised for matcher Value - Required checkbox.
```
Comment (oshtaier) [inline tests/foreman/api/test_classparameters.py]: `Error is not raised`

File: tests/foreman/api/test_classparameters.py [hunk 5]
```diff
@@ -0,0 +1,1035 @@
...
+        @assert: Error not raised for matcher value.
+
+        @status: Manual
+        """
+
+    @run_only_on('sat')
+    @stubbed()
+    @tier3
+    def test_negative_validate_default_value_with_regex(self):
+        """Error raised for default value not matching with regex.
```
Comment (oshtaier) [inline tests/foreman/api/test_classparameters.py]: Like above. Please, correct all occurences

File: tests/foreman/api/test_classparameters.py [hunk 6]
```diff
@@ -0,0 +1,1035 @@
...
+
+        @feature: Class Parameters - Optional Validation
+
+        @steps:
+
+        1.  Set override to True.
+        2.  Set default value that doesn't matches the list of step 3.
+        3.  Validate this value with list validator type and rule.
+
+        @assert: Error raised for default value not in list.
```
Comment (oshtaier) [inline tests/foreman/api/test_classparameters.py]: `Error is raised for default value that not in the list.` Please, correct all occurences

File: tests/foreman/api/test_classparameters.py [hunk 7]
```diff
@@ -0,0 +1,1035 @@
...
+        @assert: Error not raised for matcher value of default type.
+
+        @status: Manual
+        """
+
+    @run_only_on('sat')
+    @stubbed()
+    @tier3
+    def test_negative_validate_matcher_and_default_value(self):
+        """Error for invalid default and matcher value both at a time.
```
Comment (oshtaier) [inline tests/foreman/api/test_classparameters.py]: Proper verb is missed here

File: tests/foreman/api/test_classparameters.py [hunk 8]
```diff
@@ -0,0 +1,1035 @@
...
+        @assert: The matcher has been created successfully.
+
+        @status: Manual
+        """
+
+    @run_only_on('sat')
+    @stubbed()
+    @tier1
+    def test_positive_create_matcher_puppet_default_value(self):
+        """Create matcher for attribute in parameter,
```
Comment (oshtaier) [inline tests/foreman/api/test_classparameters.py]: Please, use proper formatting here (80 characters per line)

File: tests/foreman/api/test_classparameters.py [hunk 9]
```diff
@@ -0,0 +1,1035 @@
...
+        3.  Duplicate values in YAML output are removed / not displayed.
+
+        @status: Manual
+        """
+
+    @run_only_on('sat')
+    @stubbed()
+    @tier1
+    def test_negative_create_matcher_avoid_duplicate(self):
+        """Duplicates not removed as they were not really present.
```
Comment (oshtaier) [inline tests/foreman/api/test_classparameters.py]: a little bit strange test case, but probably still valid. Anyway, please use
`Duplicates are not removed as they were not really present.`

File: tests/foreman/api/test_classparameters.py [hunk 10]
```diff
@@ -0,0 +1,1035 @@
...
+    def test_positive_enable_merge_overrides_default_checkboxes(self):
+        """Enable Merge Overrides, Merge Default checkbox for supported types.
+
+        @feature: Class Parameters - Matcher
+
+        @steps:
+
+        1.  Set parameter type to array/hash.
+
+        @assert: The Merge Overrides, Merge Default checks are
```
Comment (oshtaier) [inline tests/foreman/api/test_classparameters.py]: Please, use proper formatting here

File: tests/foreman/api/test_classparameters.py [hunk 11]
```diff
@@ -0,0 +1,1035 @@
...
+    def test_negative_enable_merge_overrides_default_checkboxes(self):
+        """Disable Merge Overrides, Merge Default checkboxes for non supported types.
+
+        @feature: Class Parameters - Matcher
+
+        @steps:
+
+        1.  Set parameter type other than array/hash.
+
+        @assert: The Merge Overrides, Merge Default checks are
```
Comment (oshtaier) [inline tests/foreman/api/test_classparameters.py]: and here

File: tests/foreman/api/test_classparameters.py [hunk 12]
```diff
@@ -0,0 +1,1035 @@
...
+        """Enable Avoid duplicates checkbox for supported type- array.
+
+        @feature: Class Parameters - Matcher
+
+        @steps:
+
+        1.  Set parameter type to array.
+        2.  Set 'merge overrides' to True.
+
+        @assert: The Avoid Duplicates check is enabled to check.
```
Comment (oshtaier) [inline tests/foreman/api/test_classparameters.py]: Probably, `The Avoid Duplicates checkbox is enabled to check.`

File: tests/foreman/api/test_classparameters.py [hunk 13]
```diff
@@ -0,0 +1,1035 @@
...
+    @stubbed()
+    @tier2
+    def test_positive_remove_matcher(self):
+        """Removal of matcher from parameter.
+
+        @feature: Class Parameters - Matcher
+
+        @steps:
+
+        1.  Override the parameter and create a matcher
```
Comment (oshtaier) [inline tests/foreman/api/test_classparameters.py]: Please, use proper formatting (till 80 characters per line)

**Comment by oshtaier** [comment]
Pending comments

File: tests/foreman/api/test_classparameters.py (hunk 8 above)
Comment (jyejare) [inline tests/foreman/api/test_classparameters.py]: Already formatted properly.

Comment (oshtaier) [inline tests/foreman/api/test_classparameters.py]: Nope
```
    def test_positive_create_matcher_puppet_default_value(self):
        """Create matcher for attribute in parameter, where Value is puppet
        default value.
```
is a proper one

File: tests/foreman/api/test_classparameters.py [hunk 14]
```diff
@@ -0,0 +1,1034 @@
...
+    @stubbed()
+    @tier3
+    def test_positive_override_checkbox(self):
+        """Override the Default Parameter value.
+
+        @feature: Class Parameters - Override
+
+        @steps:
+
+        1.  Set override to True.
```
Comment (elyezer) [inline tests/foreman/api/test_classparameters.py]: Separate list index from the content with just one space. Please fix this for all the added stubs.

**Comment by elyezer** [comment]
ACK, pending nitpick.

File: tests/foreman/api/test_classparameters.py (hunk 14 above)
Comment (jyejare) [inline tests/foreman/api/test_classparameters.py]: Done,

File: tests/foreman/api/test_classparameters.py (hunk 10 above)
Comment (jyejare) [inline tests/foreman/api/test_classparameters.py]: I thinks its already formatted well. Can you please give an example of proper formatting.

Comment (oshtaier) [inline tests/foreman/api/test_classparameters.py]:
```
    @assert: The Merge Overrides, Merge Default checks are enabled to
    check.
```

**Comment by jyejare** [comment]
@elyezer and @oshtaier : All the comments has been taken care. Please take a look again.

**Comment by elyezer** [comment]
ACK

**Pull request merged**